File: sort/merge_sort/rust/merge_sort.rs
fn merge(v: &mut Vec<i32>, start: usize, mid: usize, end: usize) {
    let mut com1 = start;
    let mut com2 = mid+1;
    let mut com_aux = 0;
    let mut v_aux = vec![0; end-start+1];
    while com1 <= mid && com2 <= end {
        if v[com1] < v[com2] {
            v_aux[com_aux] = v[com1];
            com1 += 1;
        } else {
            v_aux[com_aux] = v[com2];
            com2 += 1;
        }
        com_aux += 1;
    }
    while com1 <= mid {
        v_aux[com_aux] = v[com1];
        com1 += 1;
        com_aux += 1;
    }
    while com2 <= end {
        v_aux[com_aux] = v[com2];
        com2 += 1;
        com_aux += 1;
    }
    for i in start..(end+1) {
        v[i] = v_aux[i - start];
    }
}
fn merge_sort(v: &mut Vec<i32>, start: usize, end: usize) {
    if start < end {
        let mid = (start+end) / 2;
        merge_sort(v, start, mid);
        merge_sort(v, mid+1, end);
        merge(v, start, mid, end);
    }
}
fn main() {
    let mut v = vec![8, 4, 6, 3, 2, 6, 9];
    println!("{:?}", v);
    let end = v.len()-1;
    merge_sort(&mut v, 0, end);
    println!("{:?}", v);
}

File: sort/merge_sort/dart/mergeSort.dart

import 'dart:math';
/**
 * Merge sort Algorithm from introduction to algorithm
 */

mergeSort(List lst,int start,int end){
  //divide and conquer
  if(start < end){
    int mid = (start+end)~/2;
    mergeSort(lst,start,mid);
    mergeSort(lst,mid+1,end);
    merge(lst,start,mid,end);
  }
}


merge(List lst,int start, int mid, int end){
    int pivot1 = mid - start + 1;
    int pivot2 = end - mid;

    List lower = List(pivot1+1);
    List upper = List(pivot2+1);
    //We have to fill with a dummy value, here infinite number.
    int intMax = pow(2,63-1);
    lower.fillRange(0, pivot1+1,intMax);
    upper.fillRange(0, pivot2+1,intMax);

    int i;
    int j;

    for(i = 1; i < pivot1+1;i++){
      lower[i-1] = lst[(start + i)-1];
    }
    for(j = 1; j < pivot2+1;j++){
      upper[j-1] = lst[mid + j];
    }

    i = 0;
    j = 0;

    for(int k = start; k < end;k++){
      if(lower[i] <= upper[j]) {
        lst[k] = lower[i];
        i++;
      }else {
        lst[k] = upper[j];
        j++;
      }
    }
}


main(){
  
  var rng = new Random();
  List test_lst = new List<int>.generate(12, (int index) => (index+1) * rng.nextInt(100));

  print(test_lst);

  mergeSort(test_lst, 0, test_lst.length-1);

  print(test_lst);

}